docs/source/more_types.rst

```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+when we try using it like so:
+
+.. code-block:: python
+
+    def should_return_str(x: object) -> str:
+        return unsafe_func(x)
+
+    bad_var = should_return_str(42)
+
+If we examine just the annotated types, it seems as if ``bad_var`` ought to be of
+type ``str``. But if we examine the runtime behavior, ``bad_var`` will actually be
+of type ``int``!
+
+To prevent these kinds of issues, mypy will detect and prohibit inherently unsafely
+overlapping overloads on a best-effort basis. Two variants are considered unsafely
+overlapping when both of the following are true:
+
+1. All of the arguments of the first variant are compatible with the second.
+2. The return type of the first variant is *not* compatible with (e.g. is not a
+   subtype of) the second.
+
+So in this example, the ``int`` argument in the first variant is a subtype of
+the ``object`` argument in the second, yet the ``int`` return type is a subtype of
+``str``. Both conditions are true, so mypy will correctly flag ``unsafe_func`` as
+being unsafe.
+
+However, it is unfortunately impossible to detect *all* unsafe overloads without
+crippling overloads to the point where they're completely unusable. For example,
+suppose we modify the above example so the function is overloaded against two
+completely unrelated types:
+
+.. code-block:: python
+
+    from typing import overload, Union
+
+    class A: pass
+    class B: pass
+
+    @overload
+    def func(x: A) -> int: ...
+
+    @overload
+    def func(x: B) -> str: ...
+
+    def func(x: Union[A, B]) -> Union[int, str]:
+        if isinstance(x, A):
+            return 42
+        else:
+            return "foobar"
+
+    def should_return_str(x: B) -> str:
+        return func(x)
+
+This program is fine under most normal conditions. However, if we
+create a new class that inherits from *both* A and B, we get the
+same kind of discrepency we had earlier:
+
+.. code-block:: python
+
+    class Combined(A, B): pass
+
+    # bad_var is of type 'str' according to mypy but is in reality of type 'int'.
+    bad_var = should_return_str(Combined())
+
+Mypy could prohibit this by flagging the definition of ``func``
+as being unsafe since ``A`` and ``B`` could theoretically overlap
+due to multiple inheritance. However, multiple inheritance is
+uncommon and making such a change would make it difficult
+to use overloads in a useful way. Consequently, mypy is designed to
+ignore the possibility of multiple inheritance altogether when
+checking overloads: it will *not* report any errors with the above program.
+
+Thankfully, these types of situations are relatively rare. What this
+does mean, however, is that you should exercise caution when using an
+overloaded function in scenarios where it can potentially receive values
+that are an instance of two seemingly unrelated types. For example,
+``Combined()`` is an instance of both ``A`` and ``B``; the empty list 
+``[]`` is an instance of both ``List[int]`` and ``List[str]``.
+
+
```

Comment:
It looks like an example of "...will be never matched" is missing in this section.